### PR TITLE
[TP-813] phased event contract

### PR DESCRIPTION
**Pull request by KorbinianK**
This pull request introduces a new smart contract, `PhasedEventRegister`, to manage multi-phase event registrations with role-based access control, along with a corresponding test suite. The contract allows for creating events, managing phases, and tracking user registrations, while the test suite ensures functionality and role enforcement.

### New Smart Contract: `PhasedEventRegister`

* **Event and Phase Management**: Added functionality to create events with multiple phases, open/close phases, and manage user registrations. Includes mappings for events, phase statuses, and user registrations.
* **Role-Based Access Control**: Integrated OpenZeppelin's `AccessControlUpgradeable` to manage roles (`EVENT_MANAGER_ROLE` and `DEFAULT_ADMIN_ROLE`) for secure operations.
* **Event Details and Validation**: Added methods to retrieve event details, check registration statuses, and validate event/phase IDs.

### Test Suite: `PhasedEventRegisterTest`

* **Role Enforcement**: Verified that only authorized roles (admin and event manager) can perform restricted actions like creating events or managing roles.
* **Event and Phase Operations**: Tested event creation, phase management (open/close), and user registration/unregistration functionality.
* **Edge Cases**: Added tests for invalid phase IDs and unauthorized actions to ensure proper error handling and security.

### CODE REVIEW

File: contracts/eventRegister/PhasedEventRegister.sol [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    /**
+     * @notice Registers the caller for a specific phase of an event.
+     * @param _eventId The ID of the event.
+     * @param _phaseId The phase to register for.
+     */
+    function register(uint256 _eventId, uint256 _phaseId) external {
+        _validateIds(_eventId, _phaseId);
+        require(phaseOpen[_eventId][_phaseId], "Phase closed");
```
Comment (bearni95) [inline contracts/eventRegister/PhasedEventRegister.sol]: Use defined errors and if/revert rather than require, as seen here:
https://github.com/taikoxyz/trailblazer-contracts/blob/0c5c1dc8f66fb1d68e7c05db43d62d2a70527d66/contracts/trailblazers-season-4/TrailblazersS1BadgesV9.sol#L42

File: contracts/eventRegister/PhasedEventRegister.sol [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+     * @param _eventId The ID of the event.
+     * @param _user The address to check.
+     * @return registered An array of booleans representing registration status for each phase.
+     */
+    function getRegistrationStatus(uint256 _eventId, address _user)
+        external
+        view
+        returns (bool[] memory registered)
+    {
+        require(_eventId > 0 && _eventId <= nextEventId, "Invalid event ID");
```
Comment (bearni95) [inline contracts/eventRegister/PhasedEventRegister.sol]: If/revert

File: contracts/eventRegister/PhasedEventRegister.sol [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+     * @return id Event ID.
+     * @return name Event name.
+     * @return totalPhases Total number of phases.
+     */
+    function getEvent(uint256 _eventId)
+        external
+        view
+        returns (uint256 id, string memory name, uint256 totalPhases)
+    {
+        require(_eventId > 0 && _eventId <= nextEventId, "Invalid event ID");
```
Comment (bearni95) [inline contracts/eventRegister/PhasedEventRegister.sol]: if/revert

File: contracts/eventRegister/PhasedEventRegister.sol [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+        Event memory e = events[_eventId];
+        return (e.id, e.name, e.totalPhases);
+    }
+
+    /**
+     * @dev Validates event and phase IDs.
+     */
+    function _validateIds(uint256 _eventId, uint256 _phaseId) internal view {
+        require(_eventId > 0 && _eventId <= nextEventId, "Invalid event ID");
```
Comment (bearni95) [inline contracts/eventRegister/PhasedEventRegister.sol]: if/revert

File: contracts/eventRegister/PhasedEventRegister.sol [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+        require(_phaseId > 0 && _phaseId <= events[_eventId].totalPhases, "Invalid phase ID");
```
Comment (bearni95) [inline contracts/eventRegister/PhasedEventRegister.sol]: if/revert

**Review by bearni95 — CHANGES_REQUESTED** [review]

File: contracts/eventRegister/PhasedEventRegister.sol [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+    /**
+     * @notice Registers the caller for a specific phase of an event.
+     * @param _eventId The ID of the event.
+     * @param _phaseId The phase to register for.
+     */
+    function register(uint256 _eventId, uint256 _phaseId) external {
+        _validateIds(_eventId, _phaseId);
+        if (!phaseOpen[_eventId][_phaseId]) revert PHASE_CLOSED();
+        registrations[_eventId][_phaseId][msg.sender] = block.timestamp;
```
Comment (bearni95) [inline contracts/eventRegister/PhasedEventRegister.sol]:
```suggestion
        registrations[_eventId][_phaseId][_msgSender()] = block.timestamp;
```

File: contracts/eventRegister/PhasedEventRegister.sol [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        emit Registered(msg.sender, _eventId, _phaseId);
```
Comment (bearni95) [inline contracts/eventRegister/PhasedEventRegister.sol]:
```suggestion
        emit Registered(_msgSender(), _eventId, _phaseId);
```

File: contracts/eventRegister/PhasedEventRegister.sol [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+contract PhasedEventRegister is Ownable2StepUpgradeable, AccessControlUpgradeable {
+    bytes32 public constant EVENT_MANAGER_ROLE = keccak256("EVENT_MANAGER_ROLE");
+
+    struct Event {
+        uint256 id;
+        string name;
+        uint256 totalPhases;
+    }
+
+    mapping(uint256 => Event) public events;
```
Comment (bearni95) [inline contracts/eventRegister/PhasedEventRegister.sol]: Named mapping parameters, as seen [here](https://github.com/taikoxyz/trailblazer-contracts/blob/0c5c1dc8f66fb1d68e7c05db43d62d2a70527d66/contracts/trailblazers-season-2/BadgeRecruitment.sol#L56)

```suggestion
    mapping(uint256 eventId => Event event) public events;
```

File: contracts/eventRegister/PhasedEventRegister.sol [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    mapping(uint256 => mapping(uint256 => bool)) public phaseOpen;
```
Comment (bearni95) [inline contracts/eventRegister/PhasedEventRegister.sol]: Named mapping items

File: contracts/eventRegister/PhasedEventRegister.sol [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    mapping(uint256 => mapping(uint256 => mapping(address => uint256))) public registrations;
```
Comment (bearni95) [inline contracts/eventRegister/PhasedEventRegister.sol]: Named mapping items

**Review by bearni95 — CHANGES_REQUESTED** [review]
